### ISSUE TITLE
Cant decrypt DB due to OOM

### ISSUE THREAD

**Issue by ashjas** [bpellin/keepassdroid#287]
Hello,

I have a kdbx DB with the following properties:
> - Encryption Algo: ChaCha20
> - Key Derivation func: argon2
> - Transform rounds 100
> - Mem 64MiB
> - Parallelism 8 threads.

size of DB ~ 200KB

keepassdroid cant open this DB, says db is too large for the phone.
Same DB is opened by keepass2Android just fine. Though as it does not use native argon2 implementation, it is extremely slow to open, but it surely does.

The OOM details from the log is as below:

> W zygote64: Throwing OutOfMemoryError "Failed to allocate a 1646282080 byte allocation with 12566528 free bytes and 189MB until OOM, max allowed footprint 14880424, growth limit 201326592"

I think since a non-native implementation is able to load a 200KB DB , this is a bug and needs to be addressed. Moreover, if the DB size is less, like 20-40KB i dont face the issue.
Is 200KB size too much to ask as of today's standards?

I tested this on a 4GB device, specifically, Mi A1.

Let me know of any other info.
I am attaching the DB i used to test this. just unzip to get the kdbx DB
Password is: abcdefgh

[argon2Github.zip](https://github.com/bpellin/keepassdroid/files/1998248/argon2Github.zip)

Thanks.

**Comment by Xavron** [comment]
I'm just getting an invalid db error with this one (InvalidDBException is thrown) and on Linux desktop it errors out as well.

**Comment by ashjas** [comment]
Strange! I rechecked on desktop and keepassdroid and no error. I'm using keepassxc 2.3.3 on desktop and v2.2.0.9 keepassdroid
Just to recheck,Did you unzip? And use the extracted kdbx?

**Comment by Xavron** [comment]
> Just to recheck,Did you unzip? And use the extracted kdbx?

Yes lol ;-)

> Encryption Algo: ChaCha20
> Key Derivation func: argon2
> Transform rounds 100
> Mem 64MiB
> Parallelism 8 threads.

Have created a new one using Keepass on Linux (234.6kb) and it looks like its opening fine with KeePassDroid on Android Oreo with Samsung A7

**Comment by ashjas** [comment]
Ok. I just can't comment on why there is compatibility issues.. **can you share yours?**  And btw, I checked on 3 of my family devices, and all oom'd out.

**Comment by Xavron** [comment]
> Ok. I just can't comment on why there is compatibility issues.. can you share yours? And btw, I checked on 3 of my family devices, and all oom'd out.

[ArgonNewDatabase.zip](https://github.com/bpellin/keepassdroid/files/1998473/ArgonNewDatabase.zip)
pass: 123456

On a side note your goes invalid a few seconds or so before mine finishes so mine should be on the right track

**Comment by Xavron** [comment]
So the actual error with your file is "Failed to allocate a 209715216 byte allocation with 6291456 free bytes and 125MB until OOM, max allowed footprint 9230296, growth limit 134217728"

209.7MB is a lot. Don't know what's going on there yet but 

Oh.. good to know that atleast we are consistent in errors. So mine should now open with keepass2Android after long delay though. So there has to be something to debug and address.

> And your DB open just fine.. and it takes a lot less time ~ 10sec and not 50sec as on your A7.

Now that's odd. Not sure why that would happen.

> So can you just tell me how do you create your database? which application version do you use to create the initial db?

Created with KeePass (Linux desktop) (ver 2.38) https://keepass.info/index.html

> How do you check the memory limit the app hits? are you debugging on ide?

I'm an Android app developer. Using Android Studio and the github project here. Not going to explain everything =)

I'm not sure why 2 different applications would produce a DB with a different result unless they weren't producing the same thing. So it could be mine is wrong and yours is right or vice versa.

**Comment by Xavron** [comment]
> Ah, you have a file attachment of 204800kb. That's the difference. No wonder.

(Developer reference)
Fixable with manifest addition: android:largeHeap="true"

**Comment by Xavron** [comment]
> i have a real DB thats about 140 KB with actual entries[with same encryption parameters].. and that OOMs out as well.. which resulted me in debugging this whole thing.

Yes, it looks like it will if you use attachments but not only entries as my version shows.

It seems to be the way the app works. It looks like it loads everything into memory even if you won't use it so it must request high memory usage from Android once it goes beyond normally allowed usage or not load everything at once. The latter probably won't happen here so they just need to add a flag (its a few seconds to fix and then put out an update).

**Comment by J-Jamet** [comment]
Hello, you have a very interesting discussion about memory, opening time and attachments.
I agree with Xavron's last comment, all the elements of the database are loaded into memory after the recovery of the encryption key.

Redoing the internal XML parser for data recovery with a buffer would solve these problems if the elements are loaded when needed. (Except if one element alone is very heavy, but in the case of the attachment, we should simply never load it in memory and delegate its opening to another application)

I don't think that "android: largeHeap" is a real solution, it is like enlarging the stomach of an obese who eats too much, it is not really a solution in the long run. (After if it's temporary for a case, why not)

On the other hand, for the loading time of the databases, it has nothing to do with the OOM. This is simply because your database settings are created for PCs with large memory capacity, but the phones do not have these capabilities at all. The algorithms for the key derivation functions work at max to decipher the opening key of the base, if the memory / round ratio is too high, it takes time (but it is better secured and obviously it's more efficient with native code). Here, with the big attachment, if we decrease the round and memory parameter for Argon2, the OOM problem is still there but the database starts quickly.

I'll try to make a better data parser and manage the opening of attachments in my fork, but the development may be long. :/

**Comment by Xavron** [comment]
> I don't think that "android: largeHeap" is a real solution, it is like enlarging the stomach of an obese who eats too much, it is not really a solution in the long run. (After if it's temporary for a case, why not)
> I'll try to make a better data parser and manage the opening of attachments in my fork, but the development may be long. :/

Agree in that going the extra mile is a better approach, though people will be waiting for a fix longer unless something else will mitigate it until then. In the end, they will either have to wait or not. My fix it quick and works right now, and even if its not the greatest way, Android does allow it =)

I'm happy someone has given it some thought though :+1:

**Comment by ashjas** [comment]
@J-Jamet moreover I'm curious to know how does keepass2android decrypts the db. Wild guess.. it must have been already not loading up the full db in memory, as it never fails to open,even if it's slow in decryption process due to non native implementation of argon2.

**Comment by J-Jamet** [comment]
The data analyzer must be good, you can look at the source code. I'll probably do it when implementing the new parser in java, it's not usually so complicated but it can take a long time to adapt.

**Comment by buster** [comment]
Same issue here. I began putting an openssh keypair into keepass, which works fine on keepassxc (Desktop) and also i was able to use keepass2android. 
I like keepassdroid better but this means, it became unusable for me.

**Comment by buster** [comment]
I have created a bounty for this issue, so maybe this is an inspiration for developers to fix it? :)
https://www.bountysource.com/issues/58299853-cant-decrypt-db-due-to-oom

Not sure, if there needs to be done something more on the developers side with bountysource, though...

**Comment by J-Jamet** [comment]
@buster I didn't know BountySource, it's great! ^^ So it motivated me to solve this problem of parser and I found the solution, I'm doing a fix on my fork. The problem came from the parser methods that did not handle the buffers well. 
But I can't make a direct pull request on KeePassDroid because I have too many modifications and my fork is detached from the original project, I don't know if a diff file could be enough?

**Comment by buster** [comment]
Great! Not sure how it works, but I think you will need to click "solve issue" on the bounty.. Anyway.. Hope this will be solved :)

**Comment by J-Jamet** [comment]
I finally made a pull request, the code solves the problem.
